### ISSUE TITLE
Android external display support

### ISSUE THREAD

**Issue by OpenSauce04** [Lime3DS/Lime-3DS-Emulator#351]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What feature are you suggesting?

Users have requested a feature which would allow an external display to be used to show one or both of the 3DS screens. This would provide an experience analogous to Drastic's external display support or the Wii U's DS virtual console emulation.

### Why would this feature be useful?

This would allow for a more faithful hardware experience for users on Android, as it would allow the two 3DS to be displayed on separate real-life displays.

**Comment by DroidlyMobile** [comment]
Media Projection is the feature we need how is Lime3DS rendering the 2 screens?

**Comment by DroidlyMobile** [comment]
I stared at the source code for hours today, only if I knew Kotlin I'd be able to understand how to implement what is called Media Projection so that the Top screen could be rendered on the surface via a second display or monitor.  If one of you devs could look into this that would be epic, as the Windows/Linux versions have this feature but not sure how the two screens render on the views.

**Comment by Papermanzero** [comment]
You can use the open source waylus exactly for that feature. Create a second screen open that with weylus on the phone or tablet.

**Comment by DroidlyMobile** [comment]
> You can use the open source waylus exactly for that feature. Create a second screen open that with weylus on the phone or tablet.

That's not the same thing as native media projection on Android.

**Comment by DroidlyMobile** [comment]
Any update to how the two screens render I can add media projection and push an update for this as right now I figured out how to basically copy the screens but I think it's slowing down the performance.

**Comment by DroidlyMobile** [comment]
Why would this be marked as low as the ability to play the top screen on my tv and the bottom on my device would be amazing to have this properly rendered with Media Projection on Android.  Please look into how the two screens render this would help adding this easy to implement feature.

**Comment by RageDeadworth** [comment]
Droidlymoblie, pls pls pls, push this update if you can, external display support would be amazing.
![i8h0vkuacor51](https://github.com/user-attachments/assets/3c7ba3c2-3720-4402-99f9-f1e9b7d7b2ff)
It works great on drastic.

**Comment by AbbyTheAbsol** [comment]
Someone's making a product called the DNA Duo that I want to buy, but support for dual display is limited on Android right now.
![q7wys3x0gohd1.jpeg](https://github.com/user-attachments/assets/fb1bd5ca-d3e8-48eb-a4a7-5aaf473434c3)

**Comment by DroidlyMobile** [comment]
All they need to do is figure out how it's done on Windows/Linux and implement Media Projection just like Drastic.  I personally have no clue how the two screens render or I'd modify the emulator myself.

**Comment by DavidRGriswold** [comment]
As Drastic is not open source, we cannot simply see how they do it (which wouldn't be simple anyway but at least would be a start). 

I glanced through the code and I think it is likely significantly harder to implement than you are making this out to be, since the android code is designed right now with only one emulation surface (the screen) in mind, and keeping track of a potential second surface within the android code to pass into the C++ code for the core would require more work than you seem to think it would. It isn't easy would take a lot of work and testing. If somebody passionate about this feature wants to take it on, awesome. Until then, be patient.

**Comment by DavidRGriswold** [comment]
For future reference for anybody who looks into this more, the API needed for doing this (at least with a Google Cast device) is probably https://developers.google.com/cast/docs/overview?utm_campaign=cast-remote-710&utm_source=gdbc&utm_medium=blog rather than the MediaProjection API mentioned above. MediaProjection seems to be about capturing a real device screen to mirror rather than creating a fully virtual screen on the cast-enabled device. As to supporting physical screens, that seems to be the wild west a bit, no idea how drastic does it.

**Comment by DroidlyMobile** [comment]
> For future reference for anybody who looks into this more, the API needed for doing this (at least with a Google Cast device) is probably https://developers.google.com/cast/docs/overview?utm_campaign=cast-remote-710&utm_source=gdbc&utm_medium=blog rather than the MediaProjection API mentioned above. MediaProjection seems to be about capturing a real device screen to mirror rather than creating a fully virtual screen on the cast-enabled device. As to supporting physical screens, that seems to be the wild west a bit, no idea how drastic does it.

Inccorrect David, I have seen someone do this for Android what they said they did though was rendered the screen twice using Media Projection.  I have also decompiled the .apk of drastic and found the source code that is how I know they are using Media Projection which is something I dabbled with myself to create a dual screen game that utilizes the external display and the main display.  The SDK you're linking isn't anything close to what Drastic is using, not sure how you read the other api wrong but I again decompiled Drastics source code and it clearly shows this.  I have added a snippet of this that is how I found out how Drastic is utilizing the second screen ability it wasn't a hunch or a matter of knowing how to research this stuff. 
![Screenshot 2024-10-10 153914](https://github.com/user-attachments/assets/82ae2d8c-2af2-49cc-98ed-88ed1485c2ed)

**Comment by OpenSauce04** [comment]
@DroidlyMobile not a fan of your tone, don't put down others who aren't thinking the same thing as you are. This is a collaborative environment.

**Comment by DavidRGriswold** [comment]
It's fine, my tone wasn't great either, I should have been more neutral. Anyway, I stand corrected about the API. 

The code to manage rendering surfaces and pass then to the core emulator code seems to be managed in EmulationActiviy.kt and the NativeLibrary code. If you want to take a crack at it that's where you should start.

**Comment by DroidlyMobile** [comment]
What tone?  You can figure out my tone by messaging?  I was stating the
facts but thanks 😊 my tone god keyboard warriors love to make something
out of nothing I didn't have a tone 😆

**Comment by DroidlyMobile** [comment]
> It's fine, my tone wasn't great either, I should have been more neutral. Anyway, I stand corrected about the API.
> 
> The code to manage rendering surfaces and pass then to the core emulator code seems to be managed in EmulationActiviy.kt and the NativeLibrary code. If you want to take a crack at it that's where you should start.

I wasn't bringing a tone at all, idk how anyone can get a tone out of messaging unless they start swearing or using caps.  I was stating what I did some people just don't leave the house and think everyone is out to get them so they take messages written on paper as snarky 😆

**Comment by DroidlyMobile** [comment]
So are we ever going to see this feature as it's a feature that is available on Windows/Linux so why isn't is added to Android yet which is the most popular devices right now.

**Comment by AbbyTheAbsol** [comment]
> So are we ever going to see this feature as it's a feature that is available on Windows/Linux so why isn't is added to Android yet which is the most popular devices right now.

Doing this on Android is a lot more complicated. Android doesn't natively support split windows, so apps have to use convoluted tricks to get it to work. I believe DraStic accomplishes this through doubling the screens. There's one invisible top screen and one invisible bottom screen that you can't see.

**Comment by DroidlyMobile** [comment]
No it doesn't I have already advised how Drastic uses this feature via the
Media Projection library.  If you can figure out how the implement this via
Windows/Linux in a second window all you need to do is render the second
window via Android using The surfaceview within Media Projection when a
secondary display is attached.  I don't know Kotlin or C or I'd deep dive
into this more myself, I only understand Java and Kotlin isn't something I
want to take the time to learn right now.

**Comment by DavidRGriswold** [comment]
> No it doesn't I have already advised how Drastic uses this feature via the
> Media Projection library.  If you can figure out how the implement this via
> Windows/Linux in a second window all you need to do is render the second
> window via Android using The surfaceview within Media Projection when a
> secondary display is attached.  I don't know Kotlin or C or I'd deep dive
> into this more myself, I only understand Java and Kotlin isn't something I
> want to take the time to learn right now.
> 

Nobody at Lime3DS works for you, or for anybody. It is a volunteer project. Stop demanding features. 

If you have figured out how to implement it, great! Implement it yourself and make a PR, but nobody currently working on it seems to be interested.

**Comment by DroidlyMobile** [comment]
Cool then why is there a section requesting features clown.  Get a life

**Comment by DavidRGriswold** [comment]
Pr #617 attempts at address this issue. If you are interested in testing and providing feedback, you can use the build from that Github action and give feedback. Please note this is still in very beta stage - expect crashing, and consider using a clean user directory.

**Comment by DroidlyMobile** [comment]
> Pr [#617](https://github.com/azahar-emu/azahar/pull/617) attempts at address this issue. If you are interested in testing and providing feedback, you can use the build from that Github action and give feedback. Please note this is still in very beta stage - expect crashing, and consider using a clean user directory.

Hey how do I get this version to test it?

**Comment by DavidRGriswold** [comment]
> Hey how do I get this version to test it?

https://github.com/azahar-emu/azahar/actions/runs/13930845909/artifacts/2775187777

For future reference in case of updates, you can always download the most recent version of a PR build by going into the android build from the Github actions, then choosing the Upload sub-heading, then finding the link for the uploaded file

![Image](https://github.com/user-attachments/assets/562c7379-1318-4298-90a0-737681ee9c28)

![Image](https://github.com/user-attachments/assets/cb67bc45-6b32-41d0-a5f4-006145c876cb)

**Comment by Shumakun** [comment]
I tried version with external display support and there are few problems with it, you can't change resolution while in single screen mode and there should be an option too disable stereoscopic 3d on one screen or add it only to external display

**Comment by DavidRGriswold** [comment]
> I tried version with external display support and there are few problems with it, you can't change resolution while in single screen mode and there should be an option too disable stereoscopic 3d on one screen or add it only to external display

Can you elaborate on the first one? As far as I can tell, resolution changes (e.g. using 3x or 4x rendering) apply to the external display just fine. 

I do not think adding separate 3d stereo options for both displays is reasonable at this point - it would require some significant rewriting of the code and would be a UX challenge. Perhaps if/when we address adding separate stereo options in Single Window mode on desktop, this can also be addressed, since the code could be re-used.

**Comment by Shumakun** [comment]
Yes you can change internal resolution, but you can't change render resolution in single window mode. The picture don't maintain aspect ratio.  About 3D yes it would be great if someone address that since this is how original hardware operate.

**Comment by DavidRGriswold** [comment]
I am still really confused by what you mean by "render resolution in single window mode". Can you give some photos or screenshots to make more clear what you mean? The resolution of the external display is set by the system, not azahar, and azahar should treat the external display like any other screen, fitting the layout into the provided resolution while maintaining the aspect ratio.

**Comment by nhobson99** [comment]
Been testing, and I do get some crashes, but here's my setup for those interested.

Main phone: Oneplus 8T 5G
Secondary phone: Pixel 3
Custom driver: Kimchii Turnip driver (25.1.0 R5)

Why two phones? Because it's Dual Screen, of course!

![Image](https://github.com/user-attachments/assets/0abcc580-4661-4cf1-839d-28f4a07bf301)

This could of course be made much more portable with a shorter USB cable and a 3D printed mount (or the like).

Steps:

1) Install Azahar on main phone
2) On main phone, enable "Simulate Secondary Display" at lowest resolution in developer options
3) On secondary phone, initiate screen casting from the main phone
(I tested with https://app.tangoapp.dev/, which works over USB between the two phones)
4) In Azahar, set the secondary display to the top screen, and set the bottom screen to landscape, single screen
5) Launch a game

Step 5 is where the crashes happen. I can upload logs later, but got work to do for now. Crash seems specific to using developer options, as I didn't get any crashes with an HDMI dongle.

Very cool feature, and I'm excited to see more!

<sup><sub>STMS</sub></sup>

EDIT: Quick side note, "Simulate Secondary Displays" is only needed for scrcpy. If you have an HDMI dongle or casting device like a Chromecast (Chromecast not recommended, very laggy), it isn't needed.

**Comment by DroidlyMobile** [comment]
This is why you need the DNA Duo it's a portable OLED display that connects
over USB C and won't eat your phone's battery.

**Comment by nhobson99** [comment]
> This is why you need the DNA Duo

Except I have no money, but I do have two phones :)
And also I love tinkering

**Comment by DavidGriswoldTeacher** [comment]
Oh interesting, I had not thought of using Simulate Secondary Display + scrcpy, that's a cool way to use a second phone as an external display, but it doesn't seem scrcpy natively supports android.  So You can use that tango program on an android web browser as a scrcpy receiver?  I set up my second phone as a chomecast receiver and used casting and found it actually worked just fine, but I don't play many games that are super input-sensitive.

**Comment by DavidRGriswold** [comment]
Oh interesting, I had not thought of using Simulate Secondary Display + scrcpy, that's a cool way to use a second phone as an external display, but it doesn't seem scrcpy natively supports android. So You can use that tango program on an android web browser as a scrcpy receiver? I set up my second phone as a chomecast receiver and used casting and found it actually worked just fine, but I don't play many games that are super input-sensitive.

**Comment by nhobson99** [comment]
Yup, Tango just utilizes WebUSB and creates its own ADB server and Scrcpy client. It's got a few other goodies too, like a file browser, app management, and a full ADB shell, but I digress. It requires the web browser to have WebUSB support, but most Chromium based browsers do so it's not too big of a hurdle.

Latency is minimal since everything is done over USB with hardware encode/decode, and it's even got a built-in frame limiter (like DroidlyMobile said, battery is a concern with this setup).

Idk why creating a virtual display with scrcpy (the --new-display=<resolution> option) doesn't show up as an external display to Drastic/Azahar. The secondary display developer option works though, and can be made pretty small and unobtrusive.

Any other questions, let me know. I'll have more time for tinkering later tonight.

**Comment by DavidGriswoldTeacher** [comment]
interesting. I looked into scrcpy before but the fact that the --virtual-display option actually lets you run a second app means it's not using the standard secondary display support, which either mirrors or needs to be handled by the running app; it might be hooking into something used on foldables or something, I'm not sure. Might be worth putting an issue in on the scrcpy discord asking if they have any way of adding that as a different commandline flag.

If i use `scrcpy --new-display` on my samsung phone, I get a crazy thing that kinda looks like DeX, but isn't quite DeX? I'm not sure what's happening there but it definitely doesn't work like a physical secondary display.

**Comment by DroidlyMobile** [comment]
This is using Media Projection look it up

**Comment by DavidRGriswold** [comment]
> This is using Media Projection look it up


I wrote this code. I don't need to look it up.

**Comment by DroidlyMobile** [comment]
Yeah I thought I was responding to someone that didn't understand how this
worked.  Drastic uses this that's how I know this.

**Comment by nhobson99** [comment]
[azahar_log.old.txt](https://github.com/user-attachments/files/19812941/azahar_log.old.txt)
[azahar_log.txt](https://github.com/user-attachments/files/19812942/azahar_log.txt)

Took a while to actually get things to start crashing again, but idk what changed since my last testing

**Comment by DavidRGriswold** [comment]
@nhobson99 i have made a build of this PR that adds an "Advanced MediaProjection API" option for the secondary screen - if used, it will allow scrcpy to see the azahar virtual display directly without needing to use the developer options.

https://github.com/DavidRGriswold/azahar/actions/runs/14716041126/artifacts/3024722282

I would appreciate it if you would try that build and give it some testing, since you seem keen. It definitely has some issues - sometimes it seems to hang on a restart for me, for example - but the more people mess with it and give feedback the more I can try to fix issues.

**Comment by nhobson99** [comment]
I gave it a quick test and it seems to work :)
May need a bit of a tutorial, because I wasn't understanding what to do at first. I was trying to add the --new-display flag in my scrcpy command thinking Azahar would find the new display, but the display is created by Azahar so it's not needed (and caused crashing).

I did the following to get it to work (from a terminal):

```
.\adb.exe shell appops set io.github.lime3ds.android PROJECT_MEDIA allow
```
Then launch Azahar, enable the Advanced MediaProjection setting and get into a game
```
.\scrcpy.exe --list-displays # This will tell you the display-id to use
.\scrcpy.exe --display-id=<whatever the previous command gave you>
```
And you should now see the game on the top screen, configurable with the second screen display type option.

Very cool, thank you for your work! Hope more people can test and provide feedback.

**Comment by DavidRGriswold** [comment]
Yes, sorry. You should not need the first adb step. Start Azahar, grant the permission, and then scrcpy should be a me to see the display. It works well at least at first but backgrounding and rotating etc can cause issues still...

I looked into leveraging --new-display but there's no way for the service made by scrcpy to capture another program's virtual screen OR for scrcpy to make a virtual display other apps can project to directly, so it had to be done from the Azahar side.

**Comment by nhobson99** [comment]
Ahhh, that makes sense.

I got a weird prompt when I first tried launching a game, before using ADB to grant recording privileges. Azahar wanted me to select whether to record an app or the entire screen. I wonder, is there a manifest value somewhere that would allow Azahar to select the correct media surface to bypass this screen? Theoretically, I guess it should ask the user for permanent screen recording permission. That seems to be what my ADB command does.

I'm still very new to Android development so I'm not sure myself, but it's a straw to grasp for at least.

**Comment by DavidRGriswold** [comment]
Not as far as I can tell, the documentation I have found has implied a new permission is always needed for the MediaProjection API., and it appears as that screen recording dialog (with no way to customize it, annoyingly, I assume to avoid malicious trickery)

**Comment by DroidlyMobile** [comment]
> Not as far as I can tell, the documentation I have found has implied a new permission is always needed for the MediaProjection API., and it appears as that screen recording dialog (with no way to customize it, annoyingly, I assume to avoid malicious trickery)

Has there been an update for this since the last build I never understand how to use Github even after using it myself for 10+ years lol.

**Comment by nhobson99** [comment]
After doing some more testing, the crashes I got were game-specific. I fixed Mario 3D Land by setting rendering to OpenGLES, and fixed Pokemon Moon by switching to Vulkan. I get full speed in both games with rendering resolution at 100%.

I still think using ADB to set the screen recording permission makes it more streamlined, and the screen has to be in landscape for it to mirror properly to the virtual display. But that seems like a reasonable compromise for such a niche setup.

Side note, it seems like touch screen input is not passed through from the virtual display if using the second screen as the bottom screen.
